Network stability improvements

ixje opened the pull request:
**What current issue(s) does this address, or what feature is it adding?**
Addresses #325 and #492 
There's a couple of issues with the current networking code that I'll only describe at a high level:
- nodes get stuck/stop syncing at some point
- debugging output is insufficient to investigate the reason of poor stability

**How did you solve this problem?**
- Added health checking of peers based on block/header requests vs actual responses within a given time threshold. Bad peers are disconnected from and moved into the `DEAD_ADDRS` list.
- Added monitoring of connected peers and if we reach 0 we reset and queue up peers. Address selection is sorted from "never connected to" to "recently disconnected from" so we always try to pick a fresh one.
- Added data throttling when our request speed exceeds a certain response speed threshold
- Move some threading calls into their proper threads (mainly into the reactor) to avoid unexpected behaviour.
- Improved the nodes output to give a bit more information on the clients and addresses. i.e.
```
neo> nodes
Total Connected: 2
Peer 0  /NEO:2.8.0/ -  18.222.168.189:20333 - IO   0.323453 MB in /   0.009913 MB out
Peer 1  /Neo:2.9.0/ -    47.52.168.73:20333 - IO   0.131664 MB in /   0.006687 MB out

KNOWN_ADDRS: 174
DEAD_ADDRS: 6
QUEUED_ADDRS: 3
```

- Added a lot of debug output and nicely format them to make analysis easier in the future.
- Added 2 "hidden" commands `disconnect_node <id>` and `handles` for debugging purposes. They do not show up in the `help` output of the prompt, but can be used.
- Fixed an issue where block persisting would pass, but the notificationDB fails emitting the event which caused the BlockPersist task to cancel and the node gets 'stuck'. 

**How did you make sure your solution works?**
For the past 2 weeks I've been monitoring and adjusting this daily. 

**Are there any special changes in the code that we should be aware of?**
1) @localhuman I'd say this one is specifically for you because it's notificationDB related which I think you work a lot with? Block `1231519` has one 'malformed' transaction that throws an exception in the NotificationDB when serialising a `Neo.Contract.Create` event.   

2) test-cases have not been finalised. I've skipped 2 because I wanted to get this PR out so people can start testing if they want.

**Known issues**
1) minor - `DEAD_ADDRS` can contain duplicates. This has little effect on usage or stability.
2) minor - about 1 in 200 client disconnects can have a connection that doesn't fully close (meaning the TCP socket stays open in `LISTENING` state shown by `netstat`). I'm trying to resolve this with the `Twisted` package authors. This can be a problem if it was the last node connected while having no fresh addresses to pick from. It then prevents reset. I've seen this happen twice over the course of 2 weeks of testing.
3) major - header and block persistence blocks the main thread/health monitoring threads leading to nodes timing out if processing takes too long. This is something I want to address in a separate PR because it will require extensive testing again given that I don't know what the effect will be if I simply move it into its own thread. The impact can be that we switch out healthy nodes because they time-out or exceed our response threshold. Note that this will not get us in a stuck state, there's always a check that will reset the node if all nodes gone bad.
4) don't enable `AcceptIncomingPeers` in your config file. It suffers from the same as point 3. and then some. It's on the todo list after addressing 3.

**Please check the following, if applicable:**

- [ ] Did you add any tests?
- [X] Did you run `make lint`?
- [X] Did you run `make test`?
- [X] Are you making a PR to a feature branch or development rather than master?
- [X] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

ixje commented:
I hope that some more people can try running an additional node with these changes and share your experiences before we do the final merging.
@jseagrave21 @ThomasLobker @localhuman

jseagrave21 commented:
@ixje I will start it tonight. Also, we can keep https://github.com/CityOfZion/neo-python/issues/721 open as a warning until resolved.
This looks amazing!

jseagrave21 commented:
First feedback: I see this error frequently:
```
[D 181120 04:30:11 NeoNode:454] [002][MAINTAIN][   47.52.168.73:10333] asking for more blocks 2985303 - 2985311 (8) stale count: 135 BCRLen: 8
[D 181120 04:30:11 NeoNode:642] [002][MAINTAIN][   47.52.168.73:10333][HEALTH][B] OK - response time 20.76559090614319
[E 181120 04:30:11 InvPayload:62] COULD NOT WRITE INVENTORY HASHES argument should be bytes, buffer or ASCII string, not 'NoneType'
```

jseagrave21 commented:
Second error I see regularly
```
[C 181120 05:31:15 _observer:131]
    Traceback (most recent call last):
      File "/mnt/c/users/seagrave/neo/neo-python/venv/lib/python3.7/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
        result = f(*args, **kw)
      File "/mnt/c/users/seagrave/neo/neo-python/venv/lib/python3.7/site-packages/klein/app.py", line 128, in execute_endpoint
        return endpoint_f(self._instance, *args, **kwargs)
      File "/mnt/c/users/seagrave/neo/neo-python/venv/lib/python3.7/site-packages/klein/app.py", line 227, in _f
        return _call(instance, f, request, *a, **kw)
      File "/mnt/c/users/seagrave/neo/neo-python/venv/lib/python3.7/site-packages/klein/app.py", line 50, in _call
        result = f(*args, **kwargs)
      File "/mnt/c/users/seagrave/neo/neo-python/venv/lib/python3.7/site-packages/neo_python-0.8.3.dev0-py3.7.egg/neo/api/utils.py", line 27, in wrapper
        response_data = gzip.compress(bytes(response_data, 'utf-8'), compresslevel=COMPRESS_FASTEST)
    TypeError: encoding without a string argument
```

xxedocxx commented:
Unfortunately, mine got stuck at 2985978 / 2985978
Logs:

[D 181120 18:38:05 NodeLeader:37] Failed connecting to 13.114.212.150:10333
[I 181120 18:38:05 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7b38>
[I 181120 18:38:05 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7128>
[D 181120 18:38:06 NodeLeader:37] Failed connecting to 99.238.249.87:10333
[I 181120 18:38:06 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7048>
[D 181120 18:38:06 NodeLeader:37] Failed connecting to 13.231.38.25:10333
[I 181120 18:38:06 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c3b670198>
[I 181120 18:38:06 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4ef58160>
[D 181120 18:38:07 NodeLeader:37] Failed connecting to 27.10.238.14:10333
[I 181120 18:38:07 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4367b198>
[I 181120 18:38:07 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4e6643c8>
[D 181120 18:38:08 NodeLeader:37] Failed connecting to 13.124.237.200:10333
[I 181120 18:38:08 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c509392b0>
[I 181120 18:38:08 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392c88>
[D 181120 18:38:09 NodeLeader:37] Failed connecting to 95.216.4.165:20037
[I 181120 18:38:09 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7f98>
[I 181120 18:38:09 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4ef81080>
[D 181120 18:38:10 NodeLeader:37] Failed connecting to 176.52.34.43:10333
[I 181120 18:38:10 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c50983f60>
[I 181120 18:38:10 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c5d51bd68>
[D 181120 18:38:11 NodeLeader:37] Failed connecting to 120.36.190.225:10333
[I 181120 18:38:11 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7b00>
[D 181120 18:38:11 NodeLeader:369] Peer check loop...checking [A:132 D:0 C:0 M:20 Q:20]
[D 181120 18:38:11 NodeLeader:376] Peer count is 0, allow for retries or queued connections to be established 0
[I 181120 18:38:11 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4ef58208>
[I 181120 18:38:12 threading:865] [MainNet] Block 2985978 / 2985978
[I 181120 18:38:12 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392c50>
[D 181120 18:38:13 NodeLeader:37] Failed connecting to 139.129.235.173:10333
[I 181120 18:38:13 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392c50>
[D 181120 18:38:13 NodeLeader:37] Failed connecting to 101.100.174.24:10333
[I 181120 18:38:13 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c43662e48>
[I 181120 18:38:13 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c5dd770f0>
[D 181120 18:38:14 NodeLeader:37] Failed connecting to 35.238.65.73:10333
[I 181120 18:38:14 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c3b7134a8>
[I 181120 18:38:14 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c50939748>
[I 181120 18:38:15 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392908>
[D 181120 18:38:16 NodeLeader:37] Failed connecting to 23.108.216.36:10333
[I 181120 18:38:16 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392908>
[D 181120 18:38:16 NodeLeader:37] Failed connecting to 159.100.244.198:10333
[I 181120 18:38:16 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c50983e48>
[I 181120 18:38:16 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf77b8>
[D 181120 18:38:17 NodeLeader:37] Failed connecting to 148.251.3.36:10333
[I 181120 18:38:17 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c5ddeab70>
[I 181120 18:38:17 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392160>
[D 181120 18:38:18 NodeLeader:37] Failed connecting to 137.116.52.110:10333
[I 181120 18:38:18 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c5ddeada0>
[I 181120 18:38:18 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4f8fe5f8>
[D 181120 18:38:19 NodeLeader:37] Failed connecting to 99.231.222.155:10333
[I 181120 18:38:19 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c5d51b080>
[I 181120 18:38:19 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c39392320>
[D 181120 18:38:20 NodeLeader:37] Failed connecting to 47.90.202.47:10333
[I 181120 18:38:20 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c5ddeac88>
[I 181120 18:38:20 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c393922e8>
[D 181120 18:38:21 NodeLeader:369] Peer check loop...checking [A:132 D:0 C:0 M:20 Q:20]
[D 181120 18:38:21 NodeLeader:376] Peer count is 0, allow for retries or queued connections to be established 1
[I 181120 18:38:27 threading:865] [MainNet] Block 2985978 / 2985978
[D 181120 18:38:31 NodeLeader:369] Peer check loop...checking [A:132 D:0 C:0 M:20 Q:20]
[D 181120 18:38:31 NodeLeader:376] Peer count is 0, allow for retries or queued connections to be established 2

ixje commented:
@xxedocxx at first sight what you've posted 
> [D 181120 18:38:05 NodeLeader:37] Failed connecting to 13.114.212.150:10333
[I 181120 18:38:05 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7b38>
[I 181120 18:38:05 _observer:131] Starting factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7128>
[D 181120 18:38:06 NodeLeader:37] Failed connecting to 99.238.249.87:10333
[I 181120 18:38:06 _observer:131] Stopping factory <neo.Network.NodeLeader.NeoClientFactory object at 0x7f0c4daf7048>

does not look like stuck to me, but like addresses that are either refusing connection because they maxed out on allowed peers or are not responding.
> [D 181120 18:38:31 NodeLeader:369] Peer check loop...checking [A:132 D:0 C:0 M:20 Q:20]
> [D 181120 18:38:31 NodeLeader:376] Peer count is 0, allow for retries or queued connections to be established 2

There should be a 3rd log message after the above saying:
> Peer count 0 exceeded max retries threshold, restarting...

I'll have to check how it works when I'm fully synced. To be fair thinking back at it I believe I only tried out everything while the node was in CATCHUP mode, not in MAINTAIN mode. Although I don't expect that to make much of a difference as in MAINTAIN mode it's just requesting data at a higher interval.

xxedocxx commented:
@ixje I restarted the api server right after I made the previous post and got stopped at 2986692 / 2986692 

[D 181120 23:19:41 NodeLeader:369] Peer check loop...checking [A:357 D:0 C:0 M:20 Q:20]
[D 181120 23:19:41 NodeLeader:372] Peer count 0 exceeded max retries threshold, restarting...
[D 181120 23:19:41 NodeLeader:131] stop_peer_check_loop, cancel: True
[D 181120 23:19:41 NodeLeader:133] stop_peer_check_loop, calling stop()
[D 181120 23:19:41 NodeLeader:136] stop_peer_check_loop, calling cancel()
[D 181120 23:19:41 NodeLeader:150] stop_check_bcr_loop, cancel: True
[D 181120 23:19:41 NodeLeader:238] Starting up nodeleader
[D 181120 23:19:41 NodeLeader:248] Starting up nodeleader: starting peer and mempool check loops
[D 181120 23:19:41 NodeLeader:121] start_peer_check_loop
[D 181120 23:19:41 NodeLeader:399] Queuing    31.223.172.5:10333 for new connection [in queue: 1 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   109.72.64.129:10333 for new connection [in queue: 2 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   18.182.64.200:10333 for new connection [in queue: 3 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing  101.100.174.24:10333 for new connection [in queue: 4 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   47.90.201.247:10333 for new connection [in queue: 5 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   54.169.48.131:10333 for new connection [in queue: 6 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   13.209.131.99:10333 for new connection [in queue: 7 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   47.74.134.169:10333 for new connection [in queue: 8 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing  31.193.129.231:10333 for new connection [in queue: 9 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   89.104.23.224:10333 for new connection [in queue: 10 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   52.193.25.105:10333 for new connection [in queue: 11 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   35.194.181.55:10333 for new connection [in queue: 12 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing    47.75.188.57:10333 for new connection [in queue: 13 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing 116.126.141.175:10333 for new connection [in queue: 14 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing  13.115.243.152:10333 for new connection [in queue: 15 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   47.52.132.248:10333 for new connection [in queue: 16 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing 183.156.101.149:10333 for new connection [in queue: 17 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   104.3.123.132:10333 for new connection [in queue: 18 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   47.74.242.179:10333 for new connection [in queue: 19 connected: 0 maxpeers:20]
[D 181120 23:19:41 NodeLeader:399] Queuing   116.62.132.58:10333 for new connection [in queue: 20 connected: 0 maxpeers:20]

I hope this helps.

ixje commented:
@xxedocxx can you please define what you consider "stopped"? I have a feeling our definitions differ leading to confusion (at least on my side).

What I can tell from those logs is that at some point you lost connection to all your peers. This was detected and it started queueing new connections. Then your shared log stops. Shortly after your shared logs I expect to see `new node created` followed by `connection established` if successful. 

Connection to peers can get lost for various reasons:
- they stopped running their client
- they considered us as a bad peer and disconnect us
- we disconnect them because because we consider them as a bad peer (based on performance) 

A moment of 0 connected peers is not per definition a bad thing. It's just part of a none perfect world.

W.r.t. queuing:
- Queuing just means we've lined up new addresses to start connecting to
- There's a delay between queuing and actually attempting to establish a connection
- Queuing does not guarantee that we can connect. The addresses we know about have been shared by other peers, but without any guarantee that they're valid. We might fail to connect to them because
   - they've reached their max connection limit or don't accept incoming connections
   - they've stopped running their client in the mean time

xxedocxx commented:
@ixje My bad I should've said "got stuck"
I haven't looked into the code deeply but your explain got me its logic.
I'd love to help if I can.

My node is running on AWS EC2 t2.medium

jseagrave21 commented:
@ixje my node stopped producing blocks on 2990906. Here is what the log looks like:

```
[D 181121 15:02:55 NodeLeader:369] Peer check loop...checking [A:2 D:798 C:3 M:50 Q:0]
[I 181121 15:02:55 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:55 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:02:56 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:02:56 +0000] "POST / HTTP/1.1" 200 40 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:02:56 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:56 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:02:56 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:56 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:02:57 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:02:56 +0000] "POST / HTTP/1.1" 200 599 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:02:57 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:56 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:02:57 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:57 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:02:57 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:02:57 +0000] "POST / HTTP/1.1" 200 5608 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:02:57 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:57 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:02:58 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:58 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:02:58 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:58 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:02:59 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:58 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:02:59 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:02:58 +0000] "POST / HTTP/1.1" 200 46 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:02:59 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:02:59 +0000] "OPTIONS / HTTP/1.1" 200 83 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:02:59 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:02:59 +0000] "OPTIONS / HTTP/1.1" 200 83 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:02:59 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:02:59 +0000] "OPTIONS / HTTP/1.1" 200 83 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:02:59 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:59 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:02:59 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:02:59 +0000] "POST / HTTP/1.1" 200 46 "http://switcheoexplorer.tech/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:02:59 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:02:59 +0000] "POST / HTTP/1.1" 200 5608 "http://switcheoexplorer.tech/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:02:59 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:02:59 +0000] "POST / HTTP/1.1" 200 599 "http://switcheoexplorer.tech/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:02:59 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:59 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:00 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:02:59 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:00 threading:865] [MainNet] Block 2990906 / 2990906
[I 181121 15:03:00 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:00 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:00 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:00 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:01 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:01 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:01 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:01 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:02 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:01 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:02 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:02 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:02 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:02 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:03 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:03 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:03 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:03:03 +0000] "POST / HTTP/1.1" 200 46 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:03:03 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:03 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:04 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:03 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:04 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 46 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:03:04 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:04 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:03:04 +0000] "OPTIONS / HTTP/1.1" 200 83 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:03:04 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:03:04 +0000] "OPTIONS / HTTP/1.1" 200 83 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:03:04 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:03:04 +0000] "OPTIONS / HTTP/1.1" 200 83 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:03:04 _observer:131] "35.184.122.220" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 599 "-" "Python/3.6 aiohttp/3.4.4"
[I 181121 15:03:04 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[I 181121 15:03:05 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 46 "http://switcheoexplorer.tech/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:03:05 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 5608 "http://switcheoexplorer.tech/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:03:05 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 83 "-" "Python-urllib/2.7"
[I 181121 15:03:05 _observer:131] "196.32.108.248" - - [21/Nov/2018:23:03:04 +0000] "POST / HTTP/1.1" 200 599 "http://switcheoexplorer.tech/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:63.0) Gecko/20100101 Firefox/63.0"
[I 181121 15:03:05 _observer:131] "82.196.0.91" - - [21/Nov/2018:23:03:05 +0000] "POST / HTTP/1.1" 200 166 "-" "Python-urllib/2.7"
[D 181121 15:03:05 NodeLeader:369] Peer check loop...checking [A:2 D:798 C:3 M:50 Q:0]
```

Also it is still connected to 3 peers and responding to requests:
```
{"connected": [{"address": "158.69.117.11", "port": 10333}, {"address": "137.117.155.93", "port": 10333}, {"address": "137.117.155.93", "port": 10333}]
```

I was thinking maybe we should have a blockheight checker tied to a loop and if at a predefined time (e.g. a minute) the blockheight is still the same then `NodeLeader` resets. What do you think?

ixje commented:
don't use the above yet. I've got more stuff incoming.

ixje commented:
- I identified an issue where we would queue and address that was already in the queue list. If that address would then fail to connect, then only 1 of the 2 instances would be removed. This later on would prevent new peers getting added to the queue because we think it's still full.

- Related to the above. I ended up making queue processing part of the peer check loop, instead of processing setup up a "connect to" task once when the address is added to the queue. This should prevent any addresses getting stuck in the queue.

- A second issue from which I do not know the root cause is; a new node got created, did not loose connection at any point, yet it would never reach the `ProtocolReady` state where it starts its data request tasks. I've added a check to make sure that all connected peers are running their tasks unless we're in data throttling mode.

- @jseagrave21 had a finding where block heights would not advance. This can happen if nodes that we're connected to are stuck themselves. Pulled his PR and adjusted it such that we not fully reset, but disconnected the peers and allow for trying out different addresses that we know off.

jseagrave21 commented:
Unrelated: @ixje  even though there aren't any conflicts, could you merge development or https://github.com/CityOfZion/neo-python/pull/724 with this branch to make live testing easier? (So, I don't have to manually implement #724 before testing)

jseagrave21 commented:
On the current run my node has remained at 0 peers for quite some time. I see peer_check_loop doing its job and restarting NodeLeader, but the node is not connecting any mew peers. However, blockheight is still advancing and I was wondering how that was possible?

ixje commented:
My best guess is that blocks were received, they took a long time to process, this caused nodes to timeout (because processing unfortunately is still blocking) but the height advanced.

jseagrave21 commented:
@ixje So, I figured out what happened with "getpeers" and I don't understand it. Somehow on my most recent pull from your branch, it missed this commit https://github.com/CityOfZion/neo-python/pull/723/commits/d6c35d732f9616b329c42ed2214b65422af41d7f

jseagrave21 commented:
After my node ran on the latest pull for ~500 min, peers were at 0 for quite awhile but blockheight was still advancing, usually lagging the best blockheight. Internally, every connection to an IP was failing and the traceback was from Twisted and stated the reason for connection failure was "Invalid Argument". Also, I could not access any website from my computer and had to restart my computer before my access was restored. @ixje is this related to the Twisted issue mentioned above?

jseagrave21 commented:
My node is at ~400 min and seemingly stable right now 🤞 But with only one accept and one connected peer. Is this acceptable node behavior?
```
[D 181123 23:12:58 NodeLeader:420] Peer check loop...checking [A:1 D:644 C:1 M:25 Q:0]
```
Or should we put a threshold of connected peers/time limit to go look for new peers, even if connected peers != 0?

ixje commented:
for the time being removing `reactor.suggestThreadPoolSize(30)` from `prompt.py` seems to resolve the "22: invalid argument" error. Can't say for sure because my Windows machine thought it was a good idea to just install updates unasked 😡

jseagrave21 commented:
Do you think this would also affect `np-api-server`? I haven't run `np-prompt` at all so far for my testing.

jseagrave21 commented:
My node is stuck at Block 3008831. PeerCheckLoop looks good: `Peer check loop...checking [A:2 D:5246 C:2 M:25 Q:1]`
And, BlockheightCheck is working. I have seen watched it "Restart NodeLeader" several times now, but blockheight is still not advancing. I feel like the BlockheightCheck needs to issue a hard reset for NodeLeader, as simply disconnecting all the connected peers does not seem to work. Thoughts?

ixje commented:
thanks for reporting, we'll have to keep an eye out why this is happening and if it's something we can prevent.

>I feel like the BlockheightCheck needs to issue a hard reset for NodeLeader, as simply disconnecting all the connected peers does not seem to work. Thoughts?

Bad idea, reason explained before here (bullet 2): https://github.com/ixje/neo-python/pull/2#issuecomment-440970425

jseagrave21 commented:
My node is stuck at Block 3015922. Everything appears to be running normally, including headerheight advancing. Again, disconnecting all peers does not seem to solve the problem. I see several of these errors:
```
[E 181127 02:15:00 InvPayload:62] COULD NOT WRITE INVENTORY HASHES argument should be bytes, buffer or ASCII string, not 'NoneType'
```
but they appear to be intermittent and not truly related to blockheight not advancing 🤷‍♂️

jseagrave21 commented:
@ixje the current build is great! My node has been running for about 22 hours now. I have two questions:
- Is it acceptable to run a node with only 1 peer for an extended time? I don't think we have to worry about stale blocks because of NEO block finality but I want to make sure. 
- My DEAD_ADDRS list is now at about 2500 and I am wondering if we should manage this list any differently in the background to ensure the node can run for extended times without getting bogged down potentially?

ixje commented:
Good to hear at least it seems more stable now. We don't have to worry about the size of the `DEAD_ADDRS` list. There are other lists internally that are far bigger than this

It's not necessarily bad to be connected to 1 peer if that is the only one that is stable. It might be a bug somewhere though because I run into the exact same scenario. Sounds suspicious 🤔

ixje commented:
> Is this acceptable connection behavior? (Most recent return of `getpeers`)
> 
> ```
> "connected": [{"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}]
> ```

That's definitely a bug. You should only be connected to an IP + PORT combination once. Will add this to my TODO/unittest list. Thanks!

jseagrave21 commented:
My node has now been running for >49 hours. The only problem I see now is peers are at `maxpeers` but they are all the same peer:
```
"connected": [{"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}, {"address": "47.254.43.76", "port": 10333}], "unconnected": [{"address": "197.210.55.242
```

jseagrave21 commented:
Another related oddity I found is that the same address has been filling the `DEAD_ADDRS` list. The `DEAD_ADDRS` list is now >13,000 and the majority of the addresses is "47.254.43.76".

I double checked the logic and I think it looks good. I haven't found the bug yet. One (easy-way-out) method would be to implement another check, probably on the same schedule as the `memcheck` and the `blockchaincheck` to remove any duplicates in the `DEAD_ADDRS` and `KNOWN_ADDRS` lists. Similar action to be taken for `Peers` but they would also be eliminated with `Disconnect()`

ixje commented:
Thanks for reporting 👍 . Instead of cleaning up I want to address the root cause and prevent them getting into the lists in the first place.

ixje commented:
some notes; 
- acceptance and working with incoming clients does work but it has only been tested shortly.
- **a higher `maxpeers` setting equals worse RPC server performance.** Something is still causing the reactor (of the `Twisted` package) to block for long times. Which in turns means RPC requests get no response. More nodes means more times this can occur. **To avoid confusion; `maxpeers` does not mean a limit on simultaneous RPC connections**
- a low `maxpeers` is better for syncing up. A high setting is arguably better when you're fully synced up. Why arguably? Because it depends on the quality of the node(s) you're connected to. You just need 1 good node. On the other hand more nodes means more likelihood one of the has the real latest chain state.

corck commented:
:+1: fixed my syncing problem.

ixje commented:
Thanks for testing 👍  I would appreciate it if you could let me know if you have any unexpected findings.

jseagrave21 commented:
@ixje https://github.com/CityOfZion/neo-python/pull/723/commits/b77e48911c7fdd037f1c306de372d4939da12bc8 fixed the previously reported error. My node is running smoothly, error free. I recommend a bigger beta group / merging at this point. 

@ThomasLobker have you tested this latest version? It is very good.

ThomasLobker commented:
@jseagrave21 I'm now testing with 0.8.2 (not development branch) and made some fixes on my own. Well I cannot really call it a fix since it's more a workaround. I am running a couple of nodes in Docker and the nodes are simply crashing anytime that there are no more connected peers. Docker will pick this up and restart the node. Also I've noticed that nodes are running a lot smoother once they are not in catch-up mode anymore.

ThomasLobker commented:
I have updated my private branch to the latest development branch. It's working fine, but unfortunately I'm still getting peer drops. Currently I am detecting if there are no more connected peers and shutting down the node, in order to let Docker restart the node and try again.

```[D 181218 14:59:33 NeoNode:641] [0][mode 2] IPv4Address(type='TCP', host='172.18.0.3', port=10333) - On neo Node loop error [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.defer.CancelledError'>: 
    ] 
[D 181218 14:59:33 NeoNode:641] [0][mode 2] IPv4Address(type='TCP', host='172.18.0.3', port=10333) - On neo Node loop error [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.defer.CancelledError'>: 
    ] 
[D 181218 14:59:33 NeoNode:641] [0][mode 2] IPv4Address(type='TCP', host='172.18.0.3', port=10333) - 3656659988 disconnected [Failure instance: Traceback (failure with no frames): <class 'twisted.internet.error.ConnectionLost'>: Connection to the other side was lost in a non-clean fashion.
    ]
[D 181218 14:59:33 NodeLeader:44] Dropped connection from neo-cli:10333 
[I 181218 14:59:33 NodeLeader:49] No more connected peers...```

ixje commented:
It cannot ensure that we never reach 0 peers for many reasons, the most straight forward is all connected peers are bad or simple close down. 

The main idea behind this PR is that it will not get stuck forever but can recover.

ixje commented:
just a note of warning; it can take up to several minutes. It will detect 0 peers and first attempt to queue new ones from its `KNOWN_ADDRS` list. Now, this list can contain all kinds of dead addresses that we received from other nodes but have not yet validated. This means it can hit a series of dead addresses before it finally finds one it can connect to. If they're all dead it will be recognised as well and will do a full restart after 3 confirmations. On restart it will mark all ever known addresses as valid again. The reason for doing this is because some addresses could have been bad in the past but have recovered now. They might also have new addresses they've learned over time, so we at least want to pick up those before we mark it as a bad node.

ThomasLobker commented:
Sounds good @ixje 

I would say, if we have reached the state of 0 connected peers and it has to search for new peers that are alive, then just start with the configured peers. As an infrastructure maintainer I have control over those peers and if they are down, it will be noticed and fixed.

jseagrave21 commented:
@ixje I just got my node back up. It is using the build from `b77e489`. It has been running well; however, I noticed this behavior and I was wondering if it is acceptable?
```
{"jsonrpc": "2.0", "id": "1", "result": {"connected": [{"address": "18.194.105.32", "port": 10333}, {"address": "47.104.177.205", "port": 10333}], "unconnected": [{"address": "18.194.105.32", "port": 10333}, {"address": "47.104.177.205", "port": 10333}], "bad": ...
```
Note how both "connected" addresses are also in the "unconnected" list.
